chain/client/src/sync/state.rs

```diff
@@ -993,25 +1015,78 @@ impl StateSync {
         Ok(())
     }
 
-    fn sync_shards_download_applying_status(
+    fn sync_shards_apply_complete_status(
         &mut self,
         shard_id: ShardId,
         shard_sync_download: &mut ShardSyncDownload,
         sync_hash: CryptoHash,
         chain: &mut Chain,
-        now: Utc,
+        load_memtrie_scheduler: &near_async::messaging::Sender<LoadMemtrieRequest>,
     ) -> Result<(), near_chain::Error> {
         // Keep waiting until our shard is on the list of results
         // (these are set via callback from ClientActor - both for sync and catchup).
         if let Some(result) = self.state_parts_apply_results.remove(&shard_id) {
-            match chain.set_state_finalize(shard_id, sync_hash, result) {
-                Ok(()) => {
-                    *shard_sync_download = ShardSyncDownload {
-                        downloads: vec![],
-                        status: ShardSyncStatus::StateDownloadComplete,
+            result?;
+            let epoch_id = chain.get_block_header(&sync_hash)?.epoch_id().clone();
+            let shard_uid = chain.epoch_manager.shard_id_to_uid(shard_id, &epoch_id)?;
+            let shard_state_header = chain.get_state_header(shard_id, sync_hash)?;
+            let chunk = shard_state_header.cloned_chunk();
+            let block_hash = chunk.prev_block();
+
+            // We synced shard state on top of _previous_ block for chunk in shard state header and applied state parts to
+            // flat storage. Now we can set flat head to hash of this block and create flat storage.
+            // If block_hash is equal to default - this means that we're all the way back at genesis.
+            // So we don't have to add the storage state for shard in such case.
+            // TODO(8438) - add additional test scenarios for this case.
+            if *block_hash != CryptoHash::default() {
+                chain.create_flat_storage_for_shard(shard_uid, &chunk)?;
+            }
+            // We schedule load memtrie when flat storage state (if any) is ready.
```

Comment:
what happens if create_flat_storage_for_shard above fails? since load memtrie depends on flat storage today, shall we take a look at the status of that operation (I guess it is also being checked while loading memtrie.

Reply:
The ? at the end of that call would return already, if the call returned an error. Is that what you're talking about?

Reply:
yes, I missed that, thanks.